File: javascript/word-count/word-count.js
function isCountedWord(candidate) {
  return candidate && typeof candidate !== 'function';
}

function countWords(result, word) {
  var count = 0;
  if (isCountedWord(result[word])) {
   count = result[word];
  }
  result[word] = count + 1;
  return result;
}

function words(phrase) {
  var candidates = phrase.match(/\S+/g);
  return candidates.reduce(countWords, {});
}

module.exports = words;

File: elixir/word-count/word_count.exs
defmodule Words do

  @split_on_non_alphanumeric ~r/[^[:alnum:]-]/u

  def count(sentence) do
    Regex.split(@split_on_non_alphanumeric, sentence)
      |> Enum.filter(fn w -> String.length(w) > 0 end)
      |> Enum.map(&String.downcase/1)
      |> Enum.reduce(%{}, &word_reducer/2)
  end
  
  defp word_reducer(word, word_map) do
    Map.update(word_map, word, 1, fn c -> c + 1 end)
  end

end
